### PR TITLE
Decouple code generation from Truffle artifact

### LINKED ISSUE #196: Decouple code generation from Truffle artifact.
Now that there is an initial etherscan implementation, decoupling code generation from Truffle `Artifact` JSON makes more sense.

This might also help implement #170

### PR DESCRIPTION
**Pull request by taminomara**
Fix #196, part of #512, merge after #554.

Refactor builder interface to separate code that loads artifacts, parses them and generates bindings into independent modules. Now code generation takes a reference to `Contract`, loading contracts from source is no longer builder's job.

Additionally, `Builder` was renamed to `ContractBuilder` in anticipation that we will have `ArtifactBuilder` in the future. Also `ethcontract_generate::contract` was renamed to `ethcontract_generate::generate`. I can move these changes into separate pullrequests if anyone things it's necessary.

Procedural macro will be updated later in #512.

### CODE REVIEW

**Comment by taminomara** [comment]
Note: ethcontract-generate readme needs an update, I'm working on it. Also some examples are currently broken, working on it too.

**Comment by taminomara** [comment]
Note 2: diff includes changes from #554. They will be gone after #554 is merged and this PR is rebased onto master.

File: ethcontract-derive/src/lib.rs [hunk 1]
```diff
@@ -122,30 +131,41 @@ struct ContractArgs {
...
+                Parameter::Methods(methods) => {
+                    for method in methods {
+                        builder
+                            .method_aliases
+                            .insert(method.signature, method.alias);
+                    }
+                }
+                Parameter::EventDerives(derives) => {
+                    builder.event_derives.extend(derives);
                 }
```
Comment (nlordell) [inline ethcontract-derive/src/lib.rs:164]: Much cleaner.

File: ethcontract-generate/src/generate.rs [hunk 2]
```diff
@@ -1,5 +1,3 @@
-#![deny(missing_docs)]
```
Comment (nlordell) [inline ethcontract-generate/src/generate.rs:1]: Can we put this back before merging?

Comment (taminomara) [reply ethcontract-generate/src/generate.rs:1]: I've made this module public, so this annotation is now redundant: there's one like this in the root `lib.rs` file.

Comment (nlordell) [reply ethcontract-generate/src/generate.rs:1]: Oh shoot! :see_no_evil: I missed that.

File: ethcontract-generate/src/lib.rs [hunk 3]
```diff
@@ -214,61 +157,72 @@ impl Builder {
...
 /// Type-safe contract bindings generated by a `Builder`. This type can be
 /// either written to file or into a token stream for use in a procedural macro.
 pub struct ContractBindings {
     /// The TokenStream representing the contract bindings.
-    tokens: TokenStream,
-    /// The output options used for serialization.
-    options: SerializationOptions,
+    pub tokens: TokenStream,
+
+    /// Format generated code sing locally installed copy of `rustfmt`.
```
Comment (vkgnosis) [inline ethcontract-generate/src/lib.rs]:
```suggestion
    /// Format generated code using locally installed copy of `rustfmt`.
```

**Review by vkgnosis — APPROVED** [review]

**Pull request merged**